### PR TITLE
Content Length can be trusted when the origin server returns a 304 

### PR DESCRIPTION
**Pull request by vmamidi**
ATS can trust the content length when the object is cached and origin server
returned a 304 Not Modified response.

When a cached chunked body goes stale, and the origin refreshes it with a 304,
before this change, ATS would send the object http 1.0 style -- neither with
a content-length nor as chunked. If that transfer fails midstream, clients don't
know the object failed to finished. They think they have the full object.

This allows ATS to send those stale chunked objects that are getting refreshed
with a 304 with a content-length, so that clients can tell the difference
between a completed transfer and a failed one.

### CODE REVIEW

**Comment by shinrich** [comment]
The http_with_origin autest seg faulted

```
/lib64/libpthread.so.0(+0x12080)[0x7f39f17b8080]
traffic_server(_ZN8HTTPInfo15object_size_getEv+0x1c)[0x639082]
traffic_server(_ZN12HttpTransact28handle_content_length_headerEPNS_5StateEP7HTTPHdrS3_+0x413)[0x6ed767]
traffic_server(_ZN12HttpTransact14build_responseEPNS_5StateEP7HTTPHdrS3_11HTTPVersion10HTTPStatusPKc+0x184)[0x6f2740]
traffic_server(_ZN12HttpTransact49handle_cache_operation_on_forward_server_responseEPNS_5StateE+0x1de6)[0x6e42da]
traffic_server(_ZN12HttpTransact37handle_forward_server_connection_openEPNS_5StateE+0x8d8)[0x6e1f12]
traffic_server(_ZN12HttpTransact27handle_response_from_serverEPNS_5StateE+0x2b5)[0x6e00f1]
traffic_server(_ZN12HttpTransact14HandleResponseEPNS_5StateE+0x79c)[0x6de64c]
traffic_server(_ZN6HttpSM32call_transact_and_set_next_stateEPFvPN12HttpTransact5StateEE+0x73)[0x6b846b]
traffic_server(_ZN6HttpSM17handle_api_returnEv+0x1aa)[0x6a0f46]
traffic_server(_ZN6HttpSM14do_api_calloutEv+0x40)[0x6c0c0a]
traffic_server(_ZN6HttpSM33state_read_server_response_headerEiPv+0x839)[0x6a2bf7]
traffic_server(_ZN6HttpSM12main_handlerEiPv+0x28b)[0x6a5a67]
traffic_server(_ZN12Continuation11handleEventEiPv+0xc9)[0x6183f9]
traffic_server[0x8a38f3]
traffic_server[0x8a44d9]
traffic_server(_ZN18UnixNetVConnection11net_read_ioEP10NetHandlerP7EThread+0x2b)[0x8a6611]
traffic_server(_ZN10NetHandler18process_ready_listEv+0xac)[0x89abec]
traffic_server(_ZN10NetHandler15waitForActivityEl+0x5de)[0x89b3b8]
traffic_server(_ZN7EThread15execute_regularEv+0x41b)[0x8d44a5]
traffic_server(_ZN7EThread7executeEv+0x107)[0x8d46a7]
traffic_server[0x8d3109]
```

File: proxy/http/HttpTransact.cc [hunk 1]
```diff
@@ -6622,7 +6622,8 @@ HttpTransact::handle_content_length_header(State *s, HTTPHdr *header, HTTPHdr *b
     TxnDebug("http_trans", "[handle_content_length_header] RESPONSE cont len in hdr is %" PRId64, header->get_content_length());
   } else {
     // No content length header.
-    if (s->source == SOURCE_CACHE) {
+    if (s->source == SOURCE_CACHE ||
+        (s->source == SOURCE_HTTP_ORIGIN_SERVER && s->hdr_info.server_response.status_get() == HTTP_STATUS_NOT_MODIFIED)) {
       // If there is no content-length header, we can
```
Comment (shinrich) [inline proxy/http/HttpTransact.cc]: Call to object_size_get() in this block crashes in the http_with_origin test

Comment (vmamidi) [reply proxy/http/HttpTransact.cc]: I have modified the code to check if the object exists in cache as http_with_origin_test responds with 304 for unconditional request.

**Comment by SolidWallOfCode** [comment]
Yes, it's interesting that currently the CL is trusted on a 304 even if there's no cache object successfully read. I'm pondering that.

**Comment by vmamidi** [comment]
> Yes, it's interesting that currently the CL is trusted on a 304 even if there's no cache object successfully read. I'm pondering that. Susan's crash and the subsequent fix indicates that must happen in production. But how is that handled currently?

CL wasn't trusted when ATS receives 304 response even if the object is in the cache prior to this change.  In production, we have noticed child caching truncated data when a parent sends a connection:close with no CL or chunked. 

Susan's crash happened because i didn't check if the object wasn't in the cache when ATS receives 304.   http_with_origin in AUTest tests a scenario when ATS receives unexpected 304( unconditional request).

**Comment by SolidWallOfCode** [comment]
Didn't the code at current line 6646-6651 trust the CL if the upstream responded with a 304? How are the new lines 6648-6649 different than the current 6650-6651? The condition at current 6646 look identical (less the `object_read` null check) to new 6630.

**Comment by SolidWallOfCode** [comment]
Ah, the old condition only worked for range requests. Hmmm.

**Comment by SolidWallOfCode** [comment]
Therefore for an upstream response of 304 to a single range request where `object_read` is `nullptr`, the CL is not marked as trusted when previously it was. I think that's OK, actually, because I expect that is the case for a cache miss and TS is returning the 304 to the user agent, in which case the CL is irrelevant (there's no body).

**Comment by zwoop** [comment]
Cherry-picked to 8.0.x

**Comment by zwoop** [comment]
@vmamidi Can you make a PR against the 7.1.x branch for this please?

**Pull request merged**